cluster-autoscaler/cloudprovider/gce/mig_info_provider_test.go

```diff
@@ -928,6 +929,68 @@ func TestGetMigMachineType(t *testing.T) {
 	}
 }
 
+func TestMultipleGetMigInstanceCallsLimited(t *testing.T) {
+	mig := &gceMig{
+		gceRef: GceRef{
+			Project: "project",
+			Zone:    "zone",
+			Name:    "base-instance-name",
+		},
+	}
+	instance := cloudprovider.Instance{
+		Id: "gce://project/zone/base-instance-name-abcd",
+	}
+	instanceRef, err := GceRefFromProviderId(instance.Id)
+	assert.Nil(t, err)
+	instance2 := cloudprovider.Instance{
+		Id: "gce://project/zone/base-instance-name-abcd2",
+	}
+	instanceRef2, err := GceRefFromProviderId(instance2.Id)
+	assert.Nil(t, err)
+
+	for name, tc := range map[string]struct {
+		refreshRateDuration              time.Duration
+		delayBetweenCalls                time.Duration
+		expectedCallsToFetchMigInstances int
+	}{
+		"0s refresh rate duration, refetch expected": {
+			refreshRateDuration:              0 * time.Second,
+			delayBetweenCalls:                0 * time.Second,
+			expectedCallsToFetchMigInstances: 2,
+		},
+		"5s refresh rate duration, 0.01s between calls, no refetch expected": {
+			refreshRateDuration:              5 * time.Second,
+			delayBetweenCalls:                10 * time.Millisecond,
+			expectedCallsToFetchMigInstances: 1,
+		},
+		"0.01s refresh rate duration, 0.01s between calls, refetch expected": {
+			refreshRateDuration:              10 * time.Millisecond,
+			delayBetweenCalls:                10 * time.Millisecond,
+			expectedCallsToFetchMigInstances: 2,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			cache := emptyCache()
+			cache.migs = map[GceRef]Mig{
+				mig.gceRef: mig,
+			}
+			cache.migBaseNameCache = map[GceRef]string{mig.GceRef(): "base-instance-name"}
+			callCounter := make(map[GceRef]int)
+			client := &mockAutoscalingGceClient{
+				fetchMigInstances: fetchMigInstancesWithCounter(nil, callCounter),
+			}
+			migLister := NewMigLister(cache)
+			provider := NewCachingMigInfoProvider(cache, migLister, client, mig.GceRef().Project, 1, tc.refreshRateDuration)
+			_, err = provider.GetMigForInstance(instanceRef)
+			assert.NoError(t, err)
+			time.Sleep(tc.delayBetweenCalls)
```

Comment:
Sleeps in the unit test are a bit fragile. Could you add a clock to the `cachingMigInfoProvider` and mock it in unit tests instead? That way we could avoid using sleep.

Reply:
done